rust/lance-encoding/src/encoder.rs

```diff
@@ -117,6 +132,131 @@ pub trait ArrayEncoder: std::fmt::Debug + Send + Sync {
     ) -> Result<EncodedArray>;
 }
 
+pub const MAX_MINIBLOCK_BYTES: u64 = 8 * 1024 - 4;
+pub const MAX_MINIBLOCK_VALUES: u64 = 4096;
+
+/// Page data that has been compressed into a series of chunks put into
+/// a single buffer.
+pub struct MiniBlockCompressed {
+    /// The buffer of compressed data
+    pub data: LanceBuffer,
+    /// Describes the size of each chunk
+    pub chunks: Vec<MiniBlockChunk>,
+    /// The number of values in the entire page
+    pub num_values: u64,
+}
+
+/// Describes the size of a mini-block chunk of data
+///
+/// Mini-block chunks are designed to be small (just a few disk sectors)
+/// and contain a power-of-two number of values (except for the last chunk)
+///
+/// To enforce this we limit a chunk to 4Ki values and slightly less than
+/// 8KiB of compressed data.  This means that even in the extreme case
+/// where we have 4 bytes of rep/def then we will have at most 24KiB of
+/// data (values, repetition, and definition) per mini-block.
+pub struct MiniBlockChunk {
+    // The number of bytes that make up the chunk
+    //
+    // This value must be less than or equal to 8Ki - 4 (8188)
+    pub num_bytes: u16,
+    // The log (base 2) of the number of values in the chunk.  If this is the final chunk
+    // then this should be 0 (the number of values will be calculated by subtracting the
+    // size of all other chunks from the total size of the page)
+    //
+    // For example, 1 would mean there are 2 values in the chunk and 12 would mean there
+    // are 4Ki values in the chunk.
+    //
+    // This must be <= 12 (i.e. <= 4096 values)
```

Comment:
not sure this will be a problem here:

in compression algorithms like `fastlanes bitpacking`, when the input is less than 1024 values, the compression algorithm itself will pad the input size to 1024 values. the padded values are needed to write to disk and needed for decoding

Reply:
That's fine.  Compressors are free to pad as much as they want.  The last block is allowed to have a non-power-of-two and, if it's less than 1024 values and fastlanes needs to pad then as long as it marks `num_bytes` correctly (to include the padding)